### PR TITLE
Parametric library

### PR DESCRIPTION
**Pull request by znicolaou**
I think I'm ready to merge the new ParameterizedLibrary class and example notebook with master!

Changes to the core code are pretty minor. They include
1) The new library class in feature_libraries/parameterized_library.py, which is really just a small wrapper around the `GeneralizedLibrary` for a specific use case. This case is when multiple trajectories are supplied that are governed by the same equations but with different parameter values. The parameter values are supplied as control input. It is important to construct the library features carefully in this case for PDE's, since we don't want to take spatial derivatives of the constant control parameters, so the library takes an input `feature_library` and an input `parameter_library` that act on the data and control input independently. The full library is essentially a `TensoredLibrary` of the two, but the `GeneralizedLibrary` enables different input features for the libraries.
2) I added an `exclude_libraries` option to `GeneralizedLibrary`, which enables more control. In our case, we want to include only the tensor products of the input, and not concatenate the individual libraries as well, since that would lead to repeated columns in the usual case where the `parameter_library` has a constant bias term.
3) I added a `differentiation_method` option to the `PDELibrary` and the `WeakPDELibrary`, to allow the user to change the differentiation method for the spatial derivatives. This addresses issue #221 
4) I made a small change to the default behavior for `AxesArray`, so that if no `axes` are provided, the axes are comprehended with `comprehend_axes`.

Most of the work here is in the example notebook examples/17_pattern_formation/17_pattern_formation.ipynb, which has several examples including discrete maps, ODEs, and PDEs. Many of the examples are spatio-temporal and require pretty significant memory and cpu resources, so if you want to run them, you may want to do so on a cluster. 

I'm hoping to merge this in the next week or so before we post an accompanying preprint on arxiv. Let me know if you want to see any details from the manuscript!

### CODE REVIEW

**Comment by akaptano** [comment]
Nice! I'll try to take a look this week. By the way, it would be nice to overhaul the GeneralizedLibrary notation anyways. The inputs_per_library and array to tell it which libraries to use for which variables is super confusing to use, and I get lots of questions on it. At the very least, it might be worth putting the indexing in the backend, and having the user interface be a lot easier to specify exactly what you want the libraries to do.

File: pysindy/feature_library/base.py
Comment (Jacob-Stevens-Haas) [inline pysindy/feature_library/base.py:344]: Why did the old method from the superclass not work?

File: pysindy/feature_library/base.py
Comment (Jacob-Stevens-Haas) [inline pysindy/feature_library/base.py:534]: Is this supposed to be a concatenate?

File: pysindy/feature_library/generalized_library.py
Comment (Jacob-Stevens-Haas) [inline pysindy/feature_library/generalized_library.py:243]: shorter:
```python
sum(lib.n_output_features for lib in fitted_libs if lib not in self.exclude_libs_)
```

File: pysindy/feature_library/parameterized_library.py
Comment (Jacob-Stevens-Haas) [inline pysindy/feature_library/parameterized_library.py:112]: NIT: Mild preference for an `isinstance` check rather than `hasattr`, as it obviates the need for the comment.

File: test/optimizers/test_optimizers.py
Comment (Jacob-Stevens-Haas) [inline test/optimizers/test_optimizers.py:1026]: delete commented code

File: test/optimizers/test_optimizers.py
Comment (Jacob-Stevens-Haas) [inline test/optimizers/test_optimizers.py:1042]: ibid

**Review by Jacob-Stevens-Haas — CHANGES_REQUESTED** [review]
Looks mostly good - a consideration with the `AxesArray` choice, and a few list comprehensions.

File: test/feature_library/test_feature_library.py
Comment (znicolaou) [reply test/feature_library/test_feature_library.py:717]: Yes, removing...thanks!

File: test/feature_library/test_feature_library.py
Comment (znicolaou) [reply test/feature_library/test_feature_library.py:568]: In the new PDELibrary and WeakPDELibrary, the `is_uniform` and `is_periodic` options became redundant when I added a spatial `differentiation_method`, which can be used to specify more  fine control. I will look a bit to see if we should raise a `DeprecationWarning` or something too...

File: pysindy/feature_library/base.py
Comment (znicolaou) [reply pysindy/feature_library/base.py:344]: If some `NewFeatureLibrary` class overwrites `calc_trajectory` to include some special processing necessary for a SINDy fit, then we'd want to allow `ConcatLibrary`'s and `TensoredLibrary`'s composed of `NewFeatureLibrary`'s to utilize the `calc_trajectory` of the first library. This is done, for example, in the new `ParameterizedLibrary`, where the time derivatives are rescaled if the library is weak. If we tried to concatenate two weak `ParameterLibrary`'s, we want that rescaling to still take place in the `ConcatLibrary`. I know that's a little complicated, but I hope it makes sense...

File: pysindy/feature_library/base.py
Comment (znicolaou) [reply pysindy/feature_library/base.py:534]: No, but it might be possible to better vectorize the products here. Currently, it loops through each pair of libraries, calculates the product of features (in a vectorized way), then appends each product to an increasing list `xp` of features. The final step here just converts the list back into a numpy axis cooresponding to `ax_coord`, as needed by `AxesArray`. Maybe there is a more general way to vectorize all products for an arbitrary list of terms...

File: pysindy/feature_library/generalized_library.py
Comment (znicolaou) [reply pysindy/feature_library/generalized_library.py:243]: Thanks!

File: pysindy/feature_library/parameterized_library.py
Comment (znicolaou) [reply pysindy/feature_library/parameterized_library.py:112]: Yeah, I agree, but there are also some edge cases that need debugging here though if people try to do exotic things. If the first library here is a `ConcatLibrary` consisting of `WeakPDELibrary`'s, it should still add this correction...we could create a `self.K` in the weak case inside of `ConcatLibrary` and `TensoredLibrary`, so that this would be true then.  I think I will implement that in my next push, in addition to these other things.

File: test/optimizers/test_optimizers.py
Comment (znicolaou) [reply test/optimizers/test_optimizers.py:1026]: Yes, thanks!

File: requirements-dev.txt
Comment (Jacob-Stevens-Haas) [reply requirements-dev.txt:26]: Sounds good, thanks.

File: pysindy/utils/axes.py
Comment (Jacob-Stevens-Haas) [reply pysindy/utils/axes.py:35]: In that case, would it be ok if `transform`  called `comprehend_axes`? e.g. 
```python
xs = [AxesArray(x, comprehend_axes(x)) for x in xs]
```
Any external interface in pysindy (e.g. `transform()`) could default to assuming a SINDy axes layout, but that assumption would belong to the caller, leaving `AxesArray` potentially useful in projects beyond pysindy that might have different conventions.

File: pysindy/feature_library/base.py
Comment (Jacob-Stevens-Haas) [reply pysindy/feature_library/base.py:344]: Makes perfect sense!

File: pysindy/feature_library/base.py
Comment (Jacob-Stevens-Haas) [reply pysindy/feature_library/base.py:534]: So like a `numpy.stack(axis=-1)`?

File: pysindy/feature_library/parameterized_library.py
Comment (Jacob-Stevens-Haas) [reply pysindy/feature_library/parameterized_library.py:112]: Ohhh, that makes sense, I think.  It probably is better as is, without adding a `K` attribute in to `ConcatLibrary`/`TensoredLibrary`, since that would break type stability.

File: pysindy/utils/axes.py
Comment (znicolaou) [reply pysindy/utils/axes.py:35]: Yeah, I'm added some updates to the wrapper `x_sequence_or_item` to do this for lists of trajectories, which will do this for `fit` and `transform`, and I'll revert the changes in `AxesArray`

File: pysindy/feature_library/parameterized_library.py
Comment (znicolaou) [reply pysindy/feature_library/parameterized_library.py:112]: I'm adding a recursive `has_weak` and `has_nonweak`, and also a better `get_spatial_grid` in my next push, to resolve this issue and use `isinstance`.

File: test/feature_library/test_feature_library.py
Comment (znicolaou) [reply test/feature_library/test_feature_library.py:568]: I updated the docstings in the libraries to reflect this, and did raise DeprecationWarnings if the options are provided.

**Comment by znicolaou** [comment]
Ok! I think I resolved issues from @Jacob-Stevens-Haas -- thanks! Let me know if anything else looks off.

**Comment by znicolaou** [comment]
bump @akaptano -- let me know if you want to take a look at this one before I merge with master this week

**Comment by akaptano** [comment]
Going to take a look at this today for some documentation/other checks but not going to get into the technical details (need to spend time getting the other PR + paper ready still). If I have any suggestions I'll let you know by the end of the day!

**Comment by akaptano** [comment]
My main comment so far is that the example notebook could be substantially cleaned up in terms of:
(1) more text cells describing what is happening
(2) Some of the loading, data processing could be moved to a utils.py type file so that the notebook can be smaller + we want the example notebooks to be relatively easy to understand for users. 
(3) It would be great if the notebooks could be reformatted according to pep8 + similar formatting to the current notebooks. For instance, the spacing for assignment and other operations everywhere should be ' = ' instead of '=' and so forth. 

I started to do a little bit of this in a new commit but then realized this would be a big change so commenting here instead.

**Comment by akaptano** [comment]
Other random comments:
1. Would be good to name the example notebook something more descriptive
2. It would be good to add an example to the Example 1 notebook for a tiny demonstration of this new library. It might also be good to show a quick example in the Example 1 (or other notebook) using the "exclude_libraries" and "differentiation_method" changes you mentioned. 

Anyways, this is awesome and my main comment is to clean up the notebook considerably. Let me know if you need a hand with anything, and sorry for my slowness on this!

**Comment by znicolaou** [comment]
Thanks for the feedback @akaptano! All very reasonable--I'll clean things up and organize some short examples for the other notebooks.

**Comment by akaptano** [comment]
Hey Zack, the notebook is looking much better. I think you could probably move even more functions/plotting into the utils.py file, but this has a green light for me when you're happy to merge it in.

**Comment by znicolaou** [comment]
Thanks @akaptano ! I am pushing just one more clean up and am going to merge later tonight!

**Pull request merged**
